File: Madhava Series/Knopp's Series/py/knopp.py
from decimal import Decimal
from time import perf_counter

k: int = 0
approximationSquared: float = 0
approximation: Decimal = Decimal(0)
previous: Decimal = Decimal(0)
finalAccuracy: int = 0
totalComputationTime: float = 0
iterationStartTime: float = 0
iterationEndTime: float = 0

def innerSummationLoop(num: int) -> float:
    result: float = sum(1 / (2*i + 1) for i in range(num + 1))
    return result

while True:

    iterationStartTime = perf_counter()
    approximationSquared += (pow(-1, k) / (k + 1)) * innerSummationLoop(k)
    approximation = Decimal(pow(approximationSquared, 1/2)) * 4
    iterationEndTime = perf_counter()

    print(f"\nIteration {k + 1}")
    print(f"Approximation = {approximation}")

    for i, char in enumerate("3.141592653589793238462643383279502884197169399375"): # using str(pi) instead of writing the whole string like I have done here somehow displays a different number??? idk
        if char != str(approximation)[i]:
            if i < 2: print("No accurate decimal places")
            else:
                print(f"{i - 2} correct decimal place(s)")
                finalAccuracy = i - 2
            break

    print(f"Iteration duration: {iterationEndTime - iterationStartTime}  seconds")
    totalComputationTime += iterationEndTime - iterationStartTime

    deviation: Decimal = Decimal(approximation - previous)
    if deviation == 0:
        print("Negligible deviation (terminating the program)\n")
        break
    elif deviation != approximation: print(f"Deviation from previous iteration: {deviation}") # this if-statement prevents a deviation from being shown during the first iteration.

    previous = approximation
    k += 1

print(f"\n\nComputed {finalAccuracy} correct decimal places in {totalComputationTime} seconds and {k + 1} iterations.\n")

File: Madhava Series/Knopp's Series/rs/knopp.rs
use std::time::{Duration, Instant};

fn main() {

    let mut iterator: f64 = 0.0;
    let mut approximation_squared: f64 = 0.0;
    let mut approximation: f64;
    let mut previous_approximation: f64 = 0.0;
    let mut deviation: f64;
    let mut final_accuracy: u8 = 0;
    let mut start_time: Instant;
    let mut iteration_time: Duration;
    let mut total_computation_time: Duration = Duration::from_secs(0); // Duration::from_secs(0) is just a temporary value in order to initialize the total_computation_time variable.

    loop {

        start_time = Instant::now();
        approximation_squared += (f64::powf(-1.0, iterator) / (iterator + 1.0)) * inner_summation_loop(iterator as u16);
        approximation = f64::powf(approximation_squared, 0.5) * 4.0;
        iteration_time = start_time.elapsed();

        println!("Iteration {}", iterator as u16 + 1);
        println!("Approximation = {:.51}", approximation);

        let mut i: u8 = 0;
        for c in String::from("3.141592653589793238462643383279502884197169399375105").chars() {
            if c != match approximation.to_string().chars().nth(i as usize) {
                Some(c2) => c2,
                None => char::from(0)
            } {
                if i < 2 { println!("No accurate decimal places.") }
                else {
                    println!("{} correct decimal place(s)", i - 2);
                    final_accuracy = i - 2;
                }
                break;
            }

            i += 1;
        }

        println!("Iteration duration: {:?}", end_time - start_time);
        total_computation_time += end_time - start_time;
    
        deviation = approximation - previous_approximation;
        if deviation.abs() < 1e-50 {
            println!("\nSummation converged. Terminating program...");
            break;
        }
        else { println!("Deviation from previous iteration: {:.51}\n", deviation) }

        previous_approximation = approximation;
        iterator += 1.0;

    }

    println!("Computed {} correct decimal places in {:?} and {} iterations.", final_accuracy, total_computation_time, iterator as u16 + 1);
}

fn inner_summation_loop (num: u16) -> f64 {
    let mut result: f64 = 0.0;

    for i in 0 .. num + 1 {
        result += 1.0 / (2.0 * (i as f64) + 1.0)
    }

    return result
}